### PR TITLE
Add @jupyterlab/debugger

### PR DESCRIPTION
**Pull request by afshin**
This pull request merges the JupyterLab debugger package into core.

### CODE REVIEW

File: packages/debugger/src/icons.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import { LabIcon } from '@jupyterlab/ui-components';
```
Comment (jasongrout) [inline packages/debugger/src/icons.ts:1]: Missing license

**Comment by jasongrout** [comment]
@blink1073 - I'm pretty sure the css issue we are seeing is because the codemirror styling is now in a shadow dom

**Comment by jasongrout** [comment]
@AlbertHilb, can you help us understand some of the ramifications here of the shadow dom changes in #8584? In order for the styling in this PR to apply, do we need to append a stylesheet to every instance of CodeMirror on the page?

**Comment by blink1073** [comment]
Discussion was here: https://github.com/jupyterlab/jupyterlab/pull/8584#issuecomment-647121551

**Comment by jasongrout** [comment]
https://github.com/fergald/docs/blob/master/explainers/css-shadow-parts-1.md gives some good explanation about how to style stuff inside a shadow dom from outside the shadow dom. However, we are facing a bit of a problem here about encapsulation: the debugger does explicitly use codemirror apis and assumes you have a codemirror editor - see https://github.com/jupyterlab/jupyterlab/blob/d2e65bb0170a173b67205968a5f296b993608085/packages/debugger/src/handlers/editor.ts#L272 (or any of the other 6 times in the file that the editor is cast as a Codemirror editor). This begs the question - do we still want to maintain the abstract editor interface, or are we giving up and specializing to Codemirror? There have been some similar questions on recent issues (I think the vim PR, for example)

**Comment by jasongrout** [comment]
Note that we also have a cast at https://github.com/jupyterlab/jupyterlab/blob/f0153e0258b32674c9aec106383ddf7b618cebab/packages/documentsearch/src/providers/notebooksearchprovider.ts#L76 (which assumes that a notebook editor is a codemirror editor) and several in https://github.com/jupyterlab/jupyterlab/blob/b09b8832ef5360bbd2af1070062bd25e37520522/packages/fileeditor-extension/src/commands.ts#L675

**Comment by echarles** [comment]
I wondered similar questions in https://github.com/jupyterlab/jupyterlab/pull/8706. At some point, I think it is fine to access lower level editor (like codemirror) for now, but it would be cool if we could keep the higher level abstraction so at some point we can review that and move/support other editors.

Disclaimer: I am a big fan of abstractions...

**Comment by afshin** [comment]
I think @AlbertHilb's point that in the past three years of having a stripped-down abstraction, we haven't really demonstrated that using multiple editor back-ends is a common use case is a strong argument for just picking an editor and providing the best possible experience in terms of performance, features, reasoning, etc.

**Comment by AlbertHilb** [comment]
@jasongrout: I don't know what exactly your needs are, but appending a `style` element with the relevant rules to `CodeMirrorEditor.host.shadowRoot` when you setup the editor instance and removing it when the instance is cleared, should suffice.

**Comment by AlbertHilb** [comment]
For rules that should apply to the editor node itself, use the [`:host` pseudo-class][1] or, if they should apply only when certain conditions are satisfied, use [`:host()` pseudo-class][2].

[1]: https://developer.mozilla.org/en-US/docs/Web/CSS/:host
[2]: https://developer.mozilla.org/en-US/docs/Web/CSS/:host()

**Comment by jasongrout** [comment]
> @jasongrout: I don't know what exactly your needs are, but appending a `style` element with the relevant rules to `CodeMirrorEditor.host.shadowRoot` when you setup the editor instance and removing it when the instance is cleared, should suffice.

So then it becomes:

1. figuring out when a codemirror instance is created (that we care about) and getting a chance to append that stylesheet. My understanding is that we could clone a node and the browser would do some internal optimizations so we wouldn't have tons of stylesheets, one for each codemirror.
2. in the future, perhaps we can have the codemirror expose one of these new "part" pseudoelements (supported only by very recent browsers) to be able to affect styling inside the shadow dom from outside the shadow dom

**Comment by AlbertHilb** [comment]
Hi @jasongrout.

> My understanding is that we could clone a node and the browser would do some internal optimizations so we wouldn't have tons of stylesheets, one for each codemirror.

When all browsers will have implemented Constructable Stylesheets, the recommended way to style Shadow  Dom is adopting (a reference to) a stylesheet

```
// Construct the stylesheet
const sheet = new CSSStyleSheet();

// Replace all styles synchronously:
sheet.replaceSync(rulesString);

// Adopt the stylesheet
shadowRoot.adoptedStyleSheets = [sheet];
``` 
Anyway, already now, [browsers are smart enough][1] to keep in memory just one copy of `style` elements with the same text content.

> in the future, perhaps we can have the codemirror expose one of these new "part" pseudoelements

Yes, and also the `::theme` pseudo-element. They need to be used in conjunction with custom elements, but this is not a major obstacle.

As side note, maybe, we can start using a special naming convection for css file used inside Shadow Dom, something like `file_name.shadow.css`, and configure webpack to load them as string (via `to-string-loader | css-loader`) instead of appending them in the HTML document head.

[1]: https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM#Internal_versus_external_styles

**Pull request merged**